### PR TITLE
Auth: Add anonymous users view and stats

### PR DESCRIPTION
**Pull request by eleijonmarck**
**What is this feature?**

- anonymous users users page
- add feature toggle `displayAnonymousStats `
- remove check for enterprise for `X-Grafana-Device-Id` header in request
- add anonusers/device count to stats

**Why do we need this feature?**
We want to surface the anonymous users to our users as well as show a count of them in the stats page.

**Special notes for your reviewer:**
image
image

Please check that:
- [ ] It works as expected from a user's perspective.
- [ ] If this is a pre-GA feature, it is behind a feature toggle.
- [ ] The docs are updated, and if this is a [notable improvement](https://grafana.com/docs/writers-toolkit/writing-guide/contribute-release-notes/#how-to-determine-if-content-belongs-in-a-whats-new-document), it's added to our [What's New](https://grafana.com/docs/writers-toolkit/writing-guide/contribute-release-notes/) doc.

### CODE REVIEW

File: public/app/core/services/backend_srv.ts
Comment (kalleep) [inline public/app/core/services/backend_srv.ts:164]: Why did we we remove edition check here?

File: pkg/services/anonymous/anonimpl/api/api.go
Comment (kalleep) [inline pkg/services/anonymous/anonimpl/api/api.go:70]:
```suggestion
		return response.ErrOrFallback(http.StatusInternalServerError, "Failed to list devices", err)
```

Comment (kalleep) [reply pkg/services/anonymous/anonimpl/api/api.go:70]: Right now I think the only error we could get is that the sql query faild and that should be a 500. But if we in the future have other error and are using `errutil.Error` this would correctly set the status code and response message

File: pkg/services/anonymous/anonimpl/api/api.go
Comment (kalleep) [inline pkg/services/anonymous/anonimpl/api/api.go:114]:
```suggestion
		return response.ErrOrFallback(http.StatusInternalServerError, "Failed to list devices", err)
```

Comment (kalleep) [reply pkg/services/anonymous/anonimpl/api/api.go:114]: Same as above

File: pkg/api/dtos/frontend_settings.go
Comment (kalleep) [inline pkg/api/dtos/frontend_settings.go:89]: Is this one used? I cannot see where we set this value

File: pkg/services/anonymous/service.go
Comment (kalleep) [inline pkg/services/anonymous/service.go:13]: This one is only used in api package, move it there and don't export it

File: public/app/features/admin/state/apis.tsx
Comment (kalleep) [inline public/app/features/admin/state/apis.tsx:47]: We could issue both request and then use `Promise.all` to wait for both to complete and then merge the result. We don't have to wait for the first one before we issue next request.

I would also remove `activeAnonymousUsers` from here and then perform this where it is displayed. Then the check there will be simpler

Comment (kalleep) [reply public/app/features/admin/state/apis.tsx:47]: Also wrap these in a try catch block

File: public/app/core/services/backend_srv.ts
Comment (eleijonmarck) [reply public/app/core/services/backend_srv.ts]: We had a discussion around it and decided to life it. Here is the discussion:

@danielkenlee:
> Yes we should based on a previous discussion about enabling OSS to benefit and also ease the implementation by not limiting the change to Enterprise on-prem only.

@Jguer:
>  this allows us to lift the license check on the frontend

File: pkg/services/anonymous/service.go
Comment (kalleep) [reply pkg/services/anonymous/service.go:13]: yeah you are right. But as you said lets split them up :)

File: pkg/services/anonymous/anonimpl/api/api.go
Comment (eleijonmarck) [reply pkg/services/anonymous/anonimpl/api/api.go:70]: thx for correcting and explaining ⚡

File: public/app/features/admin/state/apis.tsx
Comment (eleijonmarck) [reply public/app/features/admin/state/apis.tsx]: I did these in two requests, as i wanted redundancy as if one fails at least we display some information. In case the anon/stats fails for instance.

Comment (kalleep) [reply public/app/features/admin/state/apis.tsx]: To handle partail success https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Promise/allSettled can be used

**Review by IevaVasiljeva — COMMENTED** [review]
Great to see progress on this! Adding some comments that I had after a quick skim through the PR.

File: pkg/services/anonymous/anonimpl/api/api.go
Comment (eleijonmarck) [reply pkg/services/anonymous/anonimpl/api/api.go]: they cant but we have a design for users and we make it consistent for anonymous users to have gravatars for the UI. In the design doc, we have the sketches. If it is something you think is really worth diving into; we could explore removing it.

File: pkg/services/anonymous/anonimpl/api/api.go
Comment (eleijonmarck) [reply pkg/services/anonymous/anonimpl/api/api.go:75]: followed the convention we have for the backend when we submit them. https://github.com/grafana/grafana/blob/main/pkg/services/anonymous/anonimpl/impl.go#L62

File: pkg/services/anonymous/anonimpl/api/api.go
Comment (eleijonmarck) [reply pkg/services/anonymous/anonimpl/api/api.go:53]: this would make sense for the stats page, perhaps we could explore that. The endpoints are needed for the anonymous users tab in the `users` page. I also included the endpoints first under /api/admin/stats or closer to it but we decided to make these endpoints part of the registration of the service, together w @kalleep . any more thoughts for this @kalleep

File: public/app/features/admin/state/reducers.ts
Comment (eleijonmarck) [reply public/app/features/admin/state/reducers.ts:215]: image
i have it there for a placeholder, as we are not able to filter/search users currently. For that we need to make a query endpoint for the devices and that is not in scope currently.

I'd rather keep these as they are needed for the interactive table and the users table. If it is a must to remove, I could look into it.

File: public/app/features/admin/ServerStats.tsx
Comment (Clarity-89) [inline public/app/features/admin/ServerStats.tsx:84]: Wonder why do we hide this data of no active devices are detected?

File: public/app/features/admin/UserListAnonymousPage.tsx
Comment (Clarity-89) [inline public/app/features/admin/UserListAnonymousPage.tsx:23]: There's no filters on this page, so this code can be removed.

File: public/app/features/admin/UserListAnonymousPage.tsx
Comment (Clarity-89) [inline public/app/features/admin/UserListAnonymousPage.tsx:75]: This whole code block isn't used and can be removed.

File: public/app/features/admin/UserListAnonymousPage.tsx
Comment (Clarity-89) [inline public/app/features/admin/UserListAnonymousPage.tsx:122]: This doesn't seem to be used either.

File: public/app/types/user.ts
Comment (Clarity-89) [inline public/app/types/user.ts:140]: Can we make all these fields camelCase to follow the [FE naming convention](https://github.com/grafana/grafana/blob/b022ddeee84225f99734b47613e5008c7e75ada4/contribute/style-guides/frontend.md#use-camelcase-for)?

File: public/app/features/admin/Users/AnonUsersTable.tsx
Comment (Clarity-89) [inline public/app/features/admin/Users/AnonUsersTable.tsx:27]:
```suggestion
  return {
    browser: userAgent.split(' ')[0],
    computer: userAgent.split(' ')[1],
  };
```

File: public/app/features/admin/Users/AnonUsersTable.tsx
Comment (Clarity-89) [inline public/app/features/admin/Users/AnonUsersTable.tsx:31]: We have [this utility](https://github.com/grafana/grafana/blob/main/public/app/features/plugins/extensions/utils.tsx#L245) that could be used here.

File: public/app/features/admin/Users/AnonUsersTable.tsx
Comment (Clarity-89) [inline public/app/features/admin/Users/AnonUsersTable.tsx:49]:
```suggestion
  <Tooltip theme="info-alt" content={value} placement="top-end" interactive={true}>
    <span>
      {truncatePart(parts.browser, 10)}
      {truncatePart(parts.computer, 10)}
    </span>
  </Tooltip>
```
An extra div wrapper is no necessary here.

File: public/app/features/admin/Users/AnonUsersTable.tsx
Comment (Clarity-89) [inline public/app/features/admin/Users/AnonUsersTable.tsx:101]: If pagination is not planned for this PR, the code for it can be removed for now.

File: public/app/features/admin/state/apis.tsx
Comment (Clarity-89) [inline public/app/features/admin/state/apis.tsx:31]: Could we maybe add the anonymous stats directly to the admin stats on backend instead of having 2 separate endpoints?

File: public/app/features/admin/state/reducers.ts
Comment (Clarity-89) [reply public/app/features/admin/state/reducers.ts:215]: The filters are not necessary for the interactive table, you can remove those and all the related code from the `UserListAnonymousPage`. In fact, I'd just remove all the code not in scope for this PR, it can be added later if needed.

File: public/app/features/admin/state/reducers.ts
Comment (Clarity-89) [inline public/app/features/admin/state/reducers.ts:243]:
```suggestion
      return {
        ...state,
        devices,
        isLoading: false,
      };
    },
  },
```
You can remove all the extra fields from here as well (and from the state type).

File: pkg/services/anonymous/anonimpl/api/api.go
Comment (Clarity-89) [reply pkg/services/anonymous/anonimpl/api/api.go:53]: I'd keep it all all undef one endpoint - `/api/admin/stats` as it's much easier. For the Users page tab we could still use the `/api/admin/stats` endpoint and just get the relevant fields from there.

Comment (kalleep) [reply pkg/services/anonymous/anonimpl/api/api.go:53]: Yeah that sounds reasonable

Comment (eleijonmarck) [reply pkg/services/anonymous/anonimpl/api/api.go:53]: @Jguer and I had a discussion, and we will reintroduce the anon service into the `hs` to call it from the `/api/admin/stats`. It is a real anti-pattern to call two endpoints for one component in the frontend.

Thx for the comments

File: public/app/features/admin/ServerStats.tsx
Comment (eleijonmarck) [reply public/app/features/admin/ServerStats.tsx:84]: This is more of a check for values being available `undefined`. It is true though that this check is not needed and should be done when we are displaying the values 👍

File: pkg/services/anonymous/anonimpl/api/api.go
Comment (eleijonmarck) [reply pkg/services/anonymous/anonimpl/api/api.go]: il add this a potential improvement / confusion. as we are also saying that the devices have a "login" which is anonymous.

Comment (eleijonmarck) [reply pkg/services/anonymous/anonimpl/api/api.go]: https://github.com/grafana/identity-access-team/issues/492

File: pkg/services/anonymous/anonimpl/api/api.go
Comment (IevaVasiljeva) [reply pkg/services/anonymous/anonimpl/api/api.go:75]: There's a comment explaining why it's needed there (basically, to make tests non-flakey). Does the same apply here?

Comment (eleijonmarck) [reply pkg/services/anonymous/anonimpl/api/api.go:75]: I actually wanted it to be the same as the backend. But maybe we do not need to have exactly the same.

Comment (eleijonmarck) [reply pkg/services/anonymous/anonimpl/api/api.go:75]: i will remove this, as this might cause confusion. thx @IevaVasiljeva

File: public/app/features/admin/ServerStats.test.tsx
Comment (eleijonmarck) [inline public/app/features/admin/ServerStats.test.tsx:59]: added tests

File: pkg/api/admin.go
Comment (eleijonmarck) [inline pkg/api/admin.go:67]:
```suggestion
```

File: pkg/services/stats/models.go
Comment (eleijonmarck) [inline pkg/services/stats/models.go:107]: @Jguer i am pretty sure that this becomes a nil pointer, if this does not exists

File: public/app/core/services/backend_srv.ts
Comment (Clarity-89) [inline public/app/core/services/backend_srv.ts:100]: I wonder why this header has not been added previously, i.e. is there a reason for not adding it?

File: public/app/features/admin/ServerStats.tsx
Comment (Clarity-89) [inline public/app/features/admin/ServerStats.tsx:89]:
```suggestion
              ...(config.featureToggles.displayAnonymousStats ? [
                    { name: 'Active devices in last 30 days', value: stats.activeDevices || 0 },
                    { name: 'Active anonymous users in last 30 days', value: Math.floor((stats.activeDevices || 0) / 3) },
                  ]
                : []),
```
Maybe we could do this and always show the active device info?

File: public/app/features/admin/ServerStats.tsx
Comment (Clarity-89) [inline public/app/features/admin/ServerStats.tsx:84]: Looks like the feature toggle names has changed? Would be good to update it in the PR description.

**Review by Clarity-89 — COMMENTED** [review]
Frontend changes look good!

File: public/app/core/services/backend_srv.ts
Comment (eleijonmarck) [reply public/app/core/services/backend_srv.ts:100]: the decision was made to do it in enterprise. the decision was made to do this for opensource now. no particular reason than that

File: public/app/features/admin/ServerStats.tsx
Comment (eleijonmarck) [reply public/app/features/admin/ServerStats.tsx:89]: it is type checking that's all

File: pkg/services/stats/models.go
Comment (Jguer) [reply pkg/services/stats/models.go:107]: No, in this case you're using struct embedding, which means they become 'fields' of the Admin stats initialized to 0, which is fine

File: public/app/core/services/backend_srv.ts
Comment (Jguer) [reply public/app/core/services/backend_srv.ts:100]: There were questions if this would be useful for OSS and if it made sense to display. The final decision was yes, having an overview of your anon users in OSS is useful

**Review by Jguer — APPROVED** [review]

File: public/app/features/admin/ServerStats.tsx
Comment (ashharrison90) [inline public/app/features/admin/ServerStats.tsx:87]: @eleijonmarck i know this is behind a feature toggle and stuff, but is this intentional? doesn't feel like the number of active users will be `numActiveDevices / 3` 😅

Comment (eleijonmarck) [reply public/app/features/admin/ServerStats.tsx:87]: @ashharrison90 yea this is intentional. We count the number of users based on devices is activeDevices/3

I can send over the design doc from product. This is products decision

Comment (ashharrison90) [reply public/app/features/admin/ServerStats.tsx:87]: ...ok 😅

Comment (IevaVasiljeva) [reply public/app/features/admin/ServerStats.tsx:87]: @eleijonmarck I can see how it's pretty confusing though. Maybe we can add a comment explaining the thinking (in one of the future PRs)?

**Pull request merged**
